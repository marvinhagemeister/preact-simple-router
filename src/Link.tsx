import { h, Component } from "preact";
import { isExternal, getUrl } from "./util";

export interface Props {
  class?: string;
  target?: string;
  activeClass?: string;
  href: string;
  children?: any;
  exact?: boolean;
}

export interface State {
  current: string;
  external: boolean;
}

export default class Link extends Component<Props, State> {
  static defaultProps = {
    exact: false,
    class: "",
    activeClass: "",
  };

  public state: State;
  constructor(props: Props) {
    super(props);

    this.state = {
      current: getUrl(),
      external: isExternal(props.href),
    };
  }

  componentWillReceiveProps(props: Props) {
    if (props.href !== this.props.href) {
      this.setState({ external: isExternal(props.href) });
    }
  }

  update = () => {
    this.setState({
      current: getUrl(),
      external: isExternal(this.props.href),
    });
  };

  componentDidMount() {
    window.addEventListener("popstate", this.update);
  }

  componentWillUnmount() {
    window.removeEventListener("popstate", this.update);
  }

  onClick = (e: Event) => {
    // Only change history if the url has actually changes and the url
    // is not external
    if (this.props.href !== window.location.href && !this.state.external) {
      // Do nothing on ctrl+click and the like
      if (
        (e instanceof KeyboardEvent &&
          (e.ctrlKey || e.metaKey || e.altKey || e.shiftKey)) ||
        (e instanceof MouseEvent && e.button !== 0)
      ) {
        return;
      }

      e.preventDefault();
      window.history.pushState({}, "foo", this.props.href);
      window.dispatchEvent(new Event("popstate", { bubbles: true }));
    } else if (this.state.external) {
      window.location.href = this.props.href;
    }
  };

  render() {
    const { exact, href, children, activeClass, target } = this.props;
    const { current } = this.state;
    const css = this.props.class || "";

    const active =
      (exact && href === current) || (!exact && current.startsWith(href))
        ? " " + activeClass
        : "";

    return (
      <a
        href={href}
        onClick={this.onClick}
        class={css + active}
        target={target}
      >
        {children}
      </a>
    );
  }
}
