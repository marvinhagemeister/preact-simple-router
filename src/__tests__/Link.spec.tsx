import { h, render } from "preact";
import { mount } from "tiny-enzyme";
import Link from "../Link";
import { renderer } from "./helpers";
import { MemoryRouter } from "..";

export function fakeEvent() {
  return { button: 0, preventDefault: () => undefined };
}

describe("Link", () => {
  it("should navigate", () => {
    const spy = jest.fn();
    const router = new MemoryRouter({});
    const c = new Link({ href: "/foo", onClick: spy }, { router });
    const e = fakeEvent();
    c.onClick(e as any);

    expect(spy.mock.calls[0]).toEqual([e, "/foo/"]);
    expect(router.url).toEqual("/foo/");
  });

  it("should navigate only if url is different", () => {
    const router = new MemoryRouter({});
    const c = new Link({ href: "/" }, { router });
    c.onClick(fakeEvent() as any);
    expect(router.url).toEqual("/");
  });

  it("should not push to history if url is external", () => {
    const router = new MemoryRouter({});
    const c = new Link({ href: "https://example.com" }, { router });
    c.onClick(fakeEvent() as any);
    expect(router.url).toEqual("/");
    expect(router.stack).toEqual(["/"]);
  });

  it("should render active class", () => {
    const link = (
      <MemoryRouter>
        <Link href="/" class="foo" activeClass="bar" />
      </MemoryRouter>
    );
    expect(renderer(link)).toMatchSnapshot();

    const link2 = (
      <MemoryRouter>
        <Link href="/foo" class="foo" activeClass="bar" />
      </MemoryRouter>
    );
    expect(renderer(link2)).toMatchSnapshot();
  });

  it("should render exact active class", () => {
    const link = (
      <MemoryRouter>
        <Link href="/" class="foo" activeClass="bar" exact />
      </MemoryRouter>
    );
    expect(renderer(link)).toMatchSnapshot();

    const link2 = (
      <MemoryRouter prefix="/bob">
        <Link href="/bob/foo" class="foo" activeClass="bar" exact />
      </MemoryRouter>
    );
    expect(renderer(link2)).toMatchSnapshot();
  });
});
