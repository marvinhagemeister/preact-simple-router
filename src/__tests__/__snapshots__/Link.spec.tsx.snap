// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render active class 1`] = `
"<div>
  <a
    href=\\"/\\"
    class=\\"foo bar\\"
  >  </a>
</div>
"
`;

exports[`Link should render active class 2`] = `
"<div>
  <a
    href=\\"/foo\\"
    class=\\"foo\\"
  >  </a>
</div>
"
`;

exports[`Link should render exact active class 1`] = `
"<div>
  <a
    href=\\"/\\"
    class=\\"foo bar\\"
  >  </a>
</div>
"
`;

exports[`Link should render exact active class 2`] = `
"<div>
  <a
    href=\\"/bob/foo\\"
    class=\\"foo\\"
  >  </a>
</div>
"
`;
